### PR TITLE
Add tests for statefulset to integration tests

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/istio/issues/31148
Fixes https://github.com/istio/istio/issues/32393

### CODE REVIEW

**Comment by stevenctl** [comment]
lgtm – since we're adding a new type of test workload, can you include it in the "regular pods" check? 

https://github.com/istio/istio/blob/47f8b2f3bddc72db5a10ca9793b1aafc29b6600f/pkg/test/framework/components/echo/echotest/filters.go#L111

**Review by stevenctl — APPROVED** [review]

File: pkg/test/framework/components/echo/config.go [hunk 1]
```diff
@@ -71,6 +71,9 @@ type Config struct {
 	// Headless (k8s only) indicates that no ClusterIP should be specified.
 	Headless bool
 
+	// StatefulSet indicates that
+	StatefulSet bool
```
Comment (hzxuzhonghu) [inline pkg/test/framework/components/echo/config.go:75]: could merge with headless? Otherwise very confusing

Comment (howardjohn) [reply pkg/test/framework/components/echo/config.go:75]: I think they may be two distinct cases. You can have headless service that is not statefulset? But maybe we can/should just merge it for simplicity

Comment (stevenctl) [reply pkg/test/framework/components/echo/config.go:75]: I think they're pretty distinct cases.

Comment (stevenctl) [inline pkg/test/framework/components/echo/config.go:75]: In `common.FillInDefaults`, if `StatefulSet` is true, we should automatically set `Headless` as well.

Comment (hzxuzhonghu) [reply pkg/test/framework/components/echo/config.go:75]: Yep, something related. Normally statefulset belong to a subset of headless service

File: pilot/pkg/networking/core/v1alpha3/name_table.go [hunk 2]
```diff
@@ -95,7 +95,12 @@ func (configgen *ConfigGeneratorImpl) BuildNameTable(node *model.Proxy, push *mo
 							Namespace: svc.Attributes.Namespace,
 							Shortname: shortName,
 						}
-						out.Table[host] = nameInfo
+						if _, f := out.Table[host]; !f || instance.Endpoint.Locality.ClusterID == node.Metadata.ClusterID {
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/name_table.go:98]: Move this to L82

**Review by hzxuzhonghu — APPROVED** [review]

File: pilot/pkg/networking/core/v1alpha3/name_table.go (hunk 2 above)
Comment (stevenctl) [inline pilot/pkg/networking/core/v1alpha3/name_table.go:98]: For non-cluster local endpoints, could we also create names like `<hostname>.<clustername>.<svc>.<ns>.svc.cluster.local` (similar to [MCS](https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/1645-multi-cluster-services-api#dns) proposal, besides `cluster.local`->`clusterset.local`)?

Some basic support for multi-cluster stateful set would be nice to have if we want to promote MC to stable.

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/name_table.go:98]: I am a bit concerned about adding a new "api" when we already have a replacement planned (MCS). Maybe we should just implement the MCS one?

**Pull request merged**